Classes/Expression.lua

```diff
@@ -0,0 +1,230 @@
+	"%%=",
+	"%^=",
+	"/=",
+}
+
+
+
+
+
+-- A list of all the comparison operators. This is used to see if an action is an assignment or a comparison.
+-- For example if "x=5;y<z" was given as input then the first action is an assignment, while the second action is a comparison.
+cExpression.m_Comparisons =
+{
+	"<",
+	">",
+	"<=",
+	">=",
+	"==",
+	"!=",
+	"~=",
+}
+
+
+
+
+
+function cExpression:new(a_Formula)
+	local Obj = {}
+	
+	a_Formula = a_Formula
+	:gsub("!=", "~=") -- Lua operator for not equal is ~=
+	:gsub("&&", " and ")
+	:gsub("||", " or ")
+	
+	setmetatable(Obj, cExpression)
+	self.__index = self
+	
+	-- The string of the formula
+	Obj.m_Formula = a_Formula
+	
+	-- All the variables that that the formula can use. For example x, y and z
+	Obj.m_Parameters = {}
+	
+	-- All the variables the formula will return after executing
+	Obj.m_ReturnValues = {}
+	
+	-- A table containing predefined variables. A new one can be added using the PredefineVariable function
+	Obj.m_PredefinedConstants = {}
+	
+	return Obj
+end
+
+
+
+
+
+-- Adds a new parameter to the formula. The formula can use this in the calculation.
+-- a_Name is a string that will be the name of the parameter
+function cExpression:AddParameter(a_Name)
+	table.insert(self.m_Parameters, a_Name)
+	return self
+end
+
+
+
+
+
+-- Makes the formula return a variable when executing
+-- a_Name is the name of the variable that will be returned.
+function cExpression:AddReturnValue(a_Name)
+	table.insert(self.m_ReturnValues, a_Name)
+	return self
+end
+
+
+
+
+
+-- Adds a new constant. The formula will be able to use this in it's calculation.
+-- a_VarName is a string that will be the name of the constant.
+-- a_Value can only be a string or a number, since the environment blocks all other functions and tables.
+function cExpression:PredefineConstant(a_VarName, a_Value)
+	table.insert(self.m_PredefinedConstants, {name = a_VarName, value = a_Value})
+	return self
+end
+
+
+
+
+
+-- Creates a safe function from the formula string, the bound parameters and the predefined variables.
+function cExpression:Compile()
+	local Arguments    = table.concat(self.m_Parameters, ", ")
+	local ReturnValues = table.concat(self.m_ReturnValues, ", ")
+	
+	local PredefinedVariables = ""
+	for _, Variable in ipairs(self.m_PredefinedConstants) do
+		local Value = Variable.value
+		if (type(Value) == "string") then
+			Value = "\"" .. Value .. "\""
+		end
+		
+		PredefinedVariables = PredefinedVariables .. "local " .. Variable.name .. " = " .. Value .. "\n"
+	end
+	
+	-- The number of comparisons. This will be used to give each comparison a name (Comp<nr>)
+	local NumComparison = 1
+	
+	-- Split the formula into actions (For example in "data=5; x<y" data=5 is an action, and x<y is an action.)
+	local Actions = StringSplitAndTrim(self.m_Formula, ";")
+	
+	-- Loop through each action to check if the action is an comparison or an assignment. Handle the actions accordingly.
+	for Idx, Action in ipairs(Actions) do
```

Comment:
WHY do we need to handle actions differently? That should be in the comment, not WHAT we're doing.
`-- If an action is an assignment in a format unsupported by Lua (a += 1), convert it into a supported format (a = a + 1)`